### PR TITLE
feat(driver providers): Add BrowserStack support.

### LINKED ISSUE #1013: Investigate BrowserStack integration

### PR DESCRIPTION
**Pull request by sjelin**
Also added BrowserStack to CI

Closes #1013

### CODE REVIEW

**Comment by sjelin** [comment]
They signed the CLA back in https://github.com/angular/protractor/pull/2599

**Comment by sjelin** [comment]
They're OK with these contributions.

**Comment by sjelin** [comment]
@juliemr I'm getting ECONNREFUSED but I don't know why.  Can you take a look?

**Comment by vedharish** [comment]
Have a look at this [PR](https://github.com/sjelin/protractor/pull/2)

File: docs/referenceConf.js [hunk 1]
```diff
@@ -14,13 +14,14 @@ exports.config = {
...
   // it is testing on. This is usually done through a Selenium Server.
-  // There are four options - specify one of the following:
+  // There are five options - specify one of the following:
   //
   // 1. seleniumServerJar - to start a standalone Selenium Server locally.
   // 2. seleniumAddress - to connect to a Selenium Server which is already
   //    running.
   // 3. sauceUser/sauceKey - to use remote Selenium Servers via Sauce Labs.
-  // 4. directConnect - to connect directly to the browser Drivers.
+  // 4. bstackUser/bstackKey - to use remote Selenium Servers via BrowserStack.
```
Comment (juliemr) [inline docs/referenceConf.js]: Let's be wordy but precise and use `browserStackUser` and `browserStackKey`

File: lib/driverProviders/browserstack.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+BrowserStackDriverProvider.prototype.updateJob = function(update) {
+
+  var self = this;
+  var deferredArray = this.drivers_.map(function(driver) {
+    var deferred = q.defer();
+    driver.getSession().then(function(session) {
+      var jobStatus = update.passed ? 'completed' : 'error';
+      log.puts('BrowserStack results available at https://www.browserstack.com/automate');
+      request({
+          url: 'https://www.browserstack.com/automate/sessions/'+session.getId()+'.json',
```
Comment (juliemr) [inline lib/driverProviders/browserstack.js]: nit: add space before and after `+`

File: lib/driverProviders/browserstack.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            'Content-Type': 'application/json',
+            'Authorization': 'Basic ' + new Buffer(self.config_.bstackUser +
+                ':' + self.config_.bstackKey).toString('base64')
+          },
+          method: 'PUT',
+          form: {
+            'status': jobStatus
+          }
+        }, function(error){
+          if(error) {
```
Comment (juliemr) [inline lib/driverProviders/browserstack.js]: nit: space after `it`

File: spec/ciBStackConf.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+var env = require('./environment.js');
+
+// The main suite of Protractor tests.
+exports.config = {
+  bstackUser: process.env.BROWSER_STACK_USERNAME,
+  bstackKey: process.env.BROWSER_STACK_ACCESS_KEY,
+
+  framework: 'jasmine2',
```
Comment (juliemr) [inline spec/ciBStackConf.js]: just `jasmine` now :)

**Comment by juliemr** [comment]
For the tests, we'll also have to add a tunnel so that browserstack can access the test server on `localhost:8081`. There's an example of scripts to do that at https://github.com/angular/angular/tree/master/scripts/browserstack

**Comment by sjelin** [comment]
All comments addressed.  I couldn't get the angular team's method to work so I went with @vedharish's PR.  If you've changed your mind and decided consistency is important I'll take another look and the Angular method.

**Comment by juliemr** [comment]
Thanks, looks good!

**Comment by juliemr** [comment]
Make sure the commits are squashed and have good messages.

**Pull request merged**